include/rmm/mr/device/cuda_async_memory_resource.hpp

```diff
@@ -0,0 +1,154 @@
+namespace mr {
+
+/**
+ * @brief `device_memory_resource` derived class that uses `cudaMallocAsync`/`cudaFreeAsync` for
+ * allocation/deallocation.
+ */
+class cuda_async_memory_resource final : public device_memory_resource {
+ public:
+  /**
+   * @brief Default constructor
+   *
+   * @throws rmm::runtime_error if the CUDA version does not support `cudaMallocAsync`
+   */
+  cuda_async_memory_resource()
+  {
+#ifdef CUDA_MALLOC_ASYNC_SUPPORT
+    // Check if cudaMallocAsync Memory pool supported
+    int device{0};
+    RMM_CUDA_TRY(cudaGetDevice(&device));
+    int v{0};
+    auto e = cudaDeviceGetAttribute(&v, cudaDevAttrMemoryPoolsSupported, device);
+    RMM_EXPECTS(e == cudaSuccess && v == 1,
+                "cudaMallocAsync not supported with this CUDA driver/runtime version");
+#else
+    RMM_FAIL("cudaMallocAsync not supported");
+#endif
+  }
+
+  ~cuda_async_memory_resource()
+  {
+#ifdef CUDA_MALLOC_ASYNC_SUPPORT
+    cudaDeviceSynchronize();
+    int device{0};
+    RMM_ASSERT_CUDA_SUCCESS(cudaGetDevice(&device));
+    cudaMemPool_t pool;
+    RMM_ASSERT_CUDA_SUCCESS(cudaDeviceGetDefaultMemPool(&pool, device));
+    RMM_ASSERT_CUDA_SUCCESS(cudaMemPoolTrimTo(pool, 0));
+#endif
+  }
+
+  cuda_async_memory_resource(cuda_async_memory_resource const&) = default;
+  cuda_async_memory_resource(cuda_async_memory_resource&&)      = default;
+  cuda_async_memory_resource& operator=(cuda_async_memory_resource const&) = default;
+  cuda_async_memory_resource& operator=(cuda_async_memory_resource&&) = default;
+
+  /**
+   * @brief Query whether the resource supports use of non-null CUDA streams for
+   * allocation/deallocation. `cuda_memory_resource` does not support streams.
+   *
+   * @returns bool true
+   */
+  bool supports_streams() const noexcept override { return true; }
+
+  /**
+   * @brief Query whether the resource supports the get_mem_info API.
+   *
+   * @return true
+   */
+  bool supports_get_mem_info() const noexcept override { return true; }
+
+ private:
+  /**
+   * @brief Allocates memory of size at least `bytes` using cudaMalloc.
+   *
+   * The returned pointer has at least 256B alignment.
+   *
+   * @throws `rmm::bad_alloc` if the requested allocation could not be fulfilled
+   *
+   * @param bytes The size, in bytes, of the allocation
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* do_allocate(std::size_t bytes, rmm::cuda_stream_view stream) override
+  {
+    void* p{nullptr};
+#ifdef CUDA_MALLOC_ASYNC_SUPPORT
+    if (bytes > 0) { RMM_CUDA_TRY(cudaMallocAsync(&p, bytes, stream.value()), rmm::bad_alloc); }
+#endif
+    return p;
+  }
+
+  /**
+   * @brief Deallocate memory pointed to by \p p.
+   *
+   * @throws Nothing.
+   *
+   * @param p Pointer to be deallocated
+   */
+  void do_deallocate(void* p, std::size_t, rmm::cuda_stream_view stream) override
+  {
+#ifdef CUDA_MALLOC_ASYNC_SUPPORT
+    if (p != nullptr) { RMM_ASSERT_CUDA_SUCCESS(cudaFreeAsync(p, stream.value())); }
+#endif
```

Comment:
You'll get unused variable warnings otherwise. 
```suggestion
#ifdef CUDA_MALLOC_ASYNC_SUPPORT
    if (p != nullptr) { RMM_ASSERT_CUDA_SUCCESS(cudaFreeAsync(p, stream.value())); }
#else
   (void)p;
   (void)stream;
#endif
```